### DISCUSSION TITLE
Shouldn't "Wait time before cure" proportional settings be based on the previous layer's pixels/area?

### DISCUSSION THREAD

**Discussion by WestonHawk** [sn4k3/UVtools#471] [Ideas]
Hi there, me again. I've been playing around with the "Wait time before cure" settings. My goal is to print a box shaped case for a circuit board directly on the bed. I am using wait times to reduce blooming and elephant foot [based on the referenced blog](https://blog.honzamrazek.cz/2022/01/prints-not-sticking-to-the-build-plate-layer-separation-rough-surface-on-a-resin-printer-resin-viscosity-the-common-denominator/).

## Pixel/area proportional wait times and previous layer

I really like the idea of wait times being automatically increased or reduced based on the number of layer pixels or the area size. But I have noticed that the suggestion tool is doing this calculation based on the current layer's pixels/area, not the previous layer.

For example, layer 1 has 1,000,000 pixels and layer 2 has 100,000 pixels. We would still want the wait time before cure for layer 2 to be longer due to the amount of time it will take for layer 1 to push out the resin. Currently however the calculation for layer 2's wait time is based on its own pixels, not layer 1, so it will have a shorter wait time than it needs and blooming is likely to occur.

This is most apparent with my print because of its hollowed out box shape. The bottom of the box is one large flat surface, then once it starts at the layer that is only walls, it immediately switches to a much shorter wait time, even though the large bottom surface 0.05mm above is still taking a long time to push out the resin. Blooming from flowing resin can be an issue for layer 3 too (even if layer 3 is identical to layer 2) since there it still only 0.1mm of space between layer 1 and layer 3. It would not need as long of a wait time as layer 2 though.

Perhaps the tool could also use an option to define how many layers to "look back" and calculate an average?

I know, this wouldn't be an issue if I rotated the box 45 degrees and used supports. It would probably print quicker too. But I think it would be satisfying to get a clean print directly off the bed like an FDM printer and not have to deal with breaking off supports.

## ~~Bottom wait time layer count~~

~~When using wait times, the bottom wait times are used for the first 20 layers regardless of how many bottom layers you actually have. Having bottom wait time layers separate from bottom cure time layers is actually a good thing. You really only need 2 bottom cure layers when using long wait times, but you do need to use longer wait times through the first 1-2mm of the print due to the bed pushing out resin. With all that said, I would like the option to change how many layers it does this for so I can see if I can get away with fewer than 20 layers without blooming.~~

~~Meanwhile, both the pixel and area proportional wait time modes do not do this. The wait time "fallback" for bottom is only used for the actual bottom cure layers. So if you have 2 bottom cure layers, by layer 3 the wait times are based on the normal layer, when I would like to force a number of "bottom wait time layers" separate from bottom cure times, just like the fixed mode currently does.~~

Edit: ~~I was wrong, it does seems to be applying the bottom wait times to the first 20 layers just like in fixed mode. Still, I'd like the options to change the number of layers it does this for.~~

Edit 2: I just realized Bottom Height is already there. I am blind.

### That's it

Let me know if I have anything blatantly wrong here. Thank you again for this masterpiece!

**Comment by sn4k3** [comment]
You have an valid point here and i will improve it to look back instead.

> Perhaps the tool could also use an option to define how many layers to "look back" and calculate an average?

Interesting thought, could be or not usefull. Need to think about it. Nevertheless have it as option wont hurt anyone.

Thanks for pointing out 👍

**Reply by msspad** [reply]
In case you care how we use dynamic WaitTimeBeforeCureS, our [UVTool script](https://gist.github.com/msspad/4291f1132707ea296884620431e2b3d2) calculates WaitTimeBeforeCureS in method SettleBuildPlate(). There are 4 cases:
- "Ground effect": For the first layers, the resin of the entire area of the build plate needs to be displaced. Thus, for the first 1mm in PositionZ, WaitTimeBeforeCureS decreases from 30s to 12s.
- The more cured pixels in the previous layers, the longer it takes the build plate to settle at the target position. So linear to the number of pixel in the previous layer, WaitTimeBeforeCureS lies within 2s-12s. If more than 12% of pixels are cured in the previous layer, WaitTimeBeforeCureS is capped at 12s.
- Big changes of the number of cured pixels can cause visible layer lines. LayerIndicesWithNumPixelDropOrLeap() marks those layer.
- Sudden changes in WaitTimeBeforeCureS can also cause visible layer lines. smoothOutWaitTimes() smooths out those wait times, including the marked layers from ReadCriticalLayerIndices().

The code covers many more subtilties that we have discovered. I'm happy to explain them and also hear how to improve it further :-)

**Reply by sn4k3** [reply]
Looks like you having fun with scripting :)
Looks much more advanced than last time ;)

**Reply by msspad** [reply]
yes, UVTools is a great tool :-)
Though I think I share your opinion that the slicer software should rather take care of most of these issues...

**Reply by sn4k3** [reply]
In fact slicers should be doing most what UVtools does, it even have more data to play with...
UVtools should be an unnecessary tool. Maybe someday...

**Reply by msspad** [reply]
Exactly. For example, I apply a matt surface on the outer surface, not on the inner surface with the script. That becomes somewhat tricky with bodies with holes... It would have been much easier and more precise with the 3D body instead of layers. Maybe someday... :-)

**Reply by sn4k3** [reply]
I tried to apply your matt surface with script, but never saw it, i have to retry again...
It's easy to skip holes (inner holes) with UVtools (Will tell more on your scripting discussion)

**Reply by msspad** [reply]
matt in the script is quite specific. For example in the image below, matt is applied at the green walls and nothing at the red ones.
![image](https://user-images.githubusercontent.com/6191764/168321984-b7ff5494-29ab-4aad-8a95-deeb49affac0.png)

Also when the model starts or ends with surfaces that are parallel to build plate, it needs to be handled specially.

**Reply by msspad** [reply]
What's also different to the UVTools Corrode function is that only the most outer pixel can have a grey value, all inner pixel should stay at 255. To determine the most outer pixel, I used the centroid in the regular case.
